File: src/main/java/botmanager/frostbalance/grid/WorldMap.java
package botmanager.frostbalance.grid;

import botmanager.IOUtils;
import botmanager.frostbalance.Frostbalance;
import botmanager.frostbalance.GuildWrapper;
import botmanager.frostbalance.Influence;
import botmanager.frostbalance.OptionFlag;
import com.google.gson.*;
import net.dv8tion.jda.api.entities.Guild;

import java.io.File;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

//TODO change worldMap to split into different types based on whether it's a main map, tutorial map, or private guild map.
/**
 * A worldmap consists of a set of tiles and all data relevant to them.
 */
public class WorldMap implements Container<Tile> {

    @Deprecated
    static List<WorldMap> worldMaps = new ArrayList<>();

    @Deprecated
    public static WorldMap get(Guild guild) {
        if (Frostbalance.bot.getSettings(guild).contains(OptionFlag.MAIN)) {
            guild = null;
        }
        for (WorldMap map : worldMaps) {
            if (map.getGuild() == null) {
                if (guild == null) return map;
            } else if (map.getGuild().equals(guild)) {
                return map;
            }
        }
        WorldMap newMap = new WorldMap(guild);
        worldMaps.add(newMap);
        return newMap;
    }

    transient Influence strongestNationalClaim = new Influence(0);

    List<Tile> loadedTiles = new ArrayList<>();

    /**
     * The guild this map is tied to.
     * If null, that means this is the main/global map.
     */
    transient GuildWrapper bGuild;

    public WorldMap(Guild guild) {
        this.bGuild = Frostbalance.bot.getGuildWrapper(guild.getId());
    }

    public static Collection<WorldMap> getMaps() {
        return worldMaps;
    }


    public static WorldMap readWorld(String guildId) {

        if (guildId == null) {
            guildId = "global";
        }

        File file = new File("data/" + Frostbalance.bot.getName() + "/" + guildId + "/map.json");
        if (file.exists()) {
            GsonBuilder gsonBuilder = new GsonBuilder();
            gsonBuilder.registerTypeAdapter(TileObject.class, new TileObjectAdapter());
            Gson gson = gsonBuilder.create();
            WorldMap worldMap = gson.fromJson(IOUtils.read(file), WorldMap.class);
            if (!guildId.equals("global")) {
                worldMap.bGuild = Frostbalance.bot.getGuildWrapper(guildId);
            }
            for (Tile tile : worldMap.loadedTiles) {
                tile.map = worldMap;
                for (TileData tileData : tile.getObjects()) {
                    tileData.tile = tile;
                }
                tile.getClaimData().tile = tile;
                for (Claim claim : tile.claimData.claims) {
                    claim.claimData = tile.claimData;
                }
                System.out.println("Loaded data for tile at " + tile.getLocation() + " (" + guildId + ")");
            }
            System.out.println("Added " + worldMap.bGuild + " world map to worldMaps list.");
            worldMaps.add(worldMap);

            return worldMap;
        }
        throw new IllegalStateException("Specified file does not exist");
    }

    public static void writeWorld(Guild guild, WorldMap map) {
        String guildId;
        if (guild == null) {
            guildId = "global";
        } else {
            guildId = guild.getId();
        }

        File file = new File("data/" + Frostbalance.bot.getName() + "/" + guildId + "/map.json");
        GsonBuilder gsonBuilder = new GsonBuilder();
        gsonBuilder.registerTypeAdapter(TileObject.class, new TileObjectAdapter());
        Gson gson = gsonBuilder.create();
        IOUtils.write(file, gson.toJson(map));
    }

    /**
     * The same as get tile, but the default tile it returns will
     * *not* be cached in the grid. Do NOT use this for functions that
     * change the tile!
     * @param hex
     * @return
     */
    @Deprecated
    public Tile getTileLazy(Hex hex) {
        for (Tile tile : loadedTiles) {
            if (tile.getLocation().equals(hex)) {
                return tile;
            }
        }
        return new Tile(this, hex);
    }

    public Tile getTile(Hex hex) {
        for (Tile tile : loadedTiles) {
            if (tile.getLocation().equals(hex)) {
                return tile;
            }
        }
        Tile newTile = new Tile(this, hex);
        loadedTiles.add(newTile);
        return newTile;
    }

    public Influence getStrongestClaim() {
        updateStrongestClaim();
        return strongestNationalClaim;
    }

    protected void updateStrongestClaim() {
        strongestNationalClaim = new Influence(0);
        for (Tile tile : loadedTiles) {
            Influence nationalStrength = tile.getClaimData().getNationalStrength();
            if (nationalStrength.compareTo(strongestNationalClaim) > 0) {
                strongestNationalClaim = nationalStrength;
            }
        }
    }

    /**
     * Tests whether this map is part of the main map, shared between
     * the three main servers.
     * @return Whether this is the main map
     */
    public boolean isMainMap() {
        if (getGuild() == null) {
            return true;
        } else return false;
    }

    public boolean isTutorialMap() {
        if (!isMainMap() && bGuild.hasFlag(OptionFlag.TUTORIAL)) {
            return true;
        } else return false;
    }

    public Guild getGuild() {
        if (bGuild == null) {
            return null;
        }
        return bGuild.getGuild();
    }

    @Override
    public Tile deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        Tile tile = context.deserialize(json, Tile.class);
        tile.setParent(this);
        return tile;
    }
}

File: src/main/java/botmanager/frostbalance/grid/WorldMap.kt
package botmanager.frostbalance.grid

import botmanager.IOUtils
import botmanager.frostbalance.Frostbalance
import botmanager.frostbalance.GameNetwork
import botmanager.frostbalance.Player
import botmanager.frostbalance.grid.coordinate.Hex
import botmanager.frostbalance.resource.ResourceDepositType
import com.google.gson.GsonBuilder
import net.dv8tion.jda.api.entities.Guild
import java.io.File
import java.util.*
import java.util.stream.Collectors

/**
 * A worldmap consists of a set of tiles and all data relevant to them.
 */
class WorldMap(@Transient var gameNetwork: GameNetwork) : Containable<GameNetwork>, Container {

    val resourceDepositTypes: List<ResourceDepositType>
        get() = gameNetwork.bot.globalResources()

    @Transient
    var highestLevelClaimData: ClaimData? = null
    var loadedTiles: MutableList<Tile> = ArrayList()

    var seed: Long = Random().nextLong()

    val players: List<PlayerCharacter>
    get() = {
        var list: MutableList<PlayerCharacter> = ArrayList()
        loadedTiles.forEach{ tile -> list.addAll(tile.objects.filterIsInstance(PlayerCharacter::class.java))}
        list
    }.invoke()

    fun getCharacter(player: Player): PlayerCharacter {
        return players.firstOrNull{ character -> character.userId == player.userWrapper.id} ?: let {
            return PlayerCharacter(player.userWrapper.id, it) //automatically added to players.
        }
    }

    /**
     * Gets an instance of the tile. This will *not*
     * create a new tile internally if the tile is empty,
     * so this method should not be used if you intend to
     * modify the tile with this method!
     * @param hex The location of the tile to render
     * @return
     */
    fun getRenderTile(hex: Hex): Tile {
        for (tile in loadedTiles) {
            if (tile.getLocation() == hex) {
                return tile
            }
        }
        return Tile(this, hex)
    }

    fun getTile(hex: Hex): Tile {
        for (tile in loadedTiles) {
            if (tile.getLocation() == hex) {
                return tile
            }
        }
        val newTile = Tile(this, hex)
        loadedTiles.add(newTile)
        return newTile
    }

    private fun clearEmptyTiles() {
        loadedTiles = loadedTiles.stream().filter { tile: Tile -> !tile.isEmpty }.collect(Collectors.toList())
    }

    val highestLevelClaim: ClaimData?
        get() {
            updateHighestLevelClaim()
            return highestLevelClaimData
        }

    fun updateHighestLevelClaim() {
        highestLevelClaimData = loadedTiles.maxByOrNull { tile -> tile.claimData.claimLevel }
                ?.claimData
    }

    override fun adopt() {
        for (tile in loadedTiles) {
            tile.setParent(this)
            tile.adopt()
        }
    }

    fun isEmpty(): Boolean {
        clearEmptyTiles()
        return loadedTiles.isEmpty()
    }

    companion object {
        @Deprecated("")
        var worldMaps: MutableList<WorldMap> = ArrayList()

        @JvmStatic
        @Deprecated("")
        operator fun get(guild: Guild): WorldMap {
            return Frostbalance.bot.getGuildWrapper(guild.id).gameNetwork.worldMap
        }

        @Deprecated("")
        fun readWorldLegacy(originalGuildId: String?) {
            var effectiveGuildId: String = originalGuildId ?: "global"
            var networkName: String = originalGuildId ?: "main"
            println("New guild name: $effectiveGuildId")
            val file = File("data/" + Frostbalance.bot.name + "/$effectiveGuildId/map.json")
            if (file.exists()) {
                val gsonBuilder = GsonBuilder()
                gsonBuilder.registerTypeAdapter(TileObject::class.java, TileObjectAdapter())
                val gson = gsonBuilder.create()
                val worldMap = gson.fromJson(IOUtils.read(file), WorldMap::class.java)
                Frostbalance.bot.getGameNetwork(networkName).worldMap = worldMap
                Frostbalance.bot.getGameNetwork(networkName).adopt()
                for (tile in worldMap.loadedTiles) {
                    tile.map = worldMap
                    for (tileData in tile.getObjects()) {
                        tileData.tile = tile
                    }
                    tile.getClaimData().tile = tile
                    for (claim in tile.claimData.claims) {
                        claim.claimData = tile.claimData
                    }
                    println("Loaded data for tile at " + tile.getLocation() + " (" + effectiveGuildId + ")")
                }
                println("Added $networkName world map to worldMaps list.")
                worldMaps.add(worldMap)
            } else {
                throw IllegalStateException("Specified file does not exist")
            }
        }

        @Deprecated("")
        fun writeWorld(guild: Guild?, map: WorldMap) {
            val guildId: String
            guildId = guild?.id ?: "global"
            map.clearEmptyTiles()
            val file = File("data/" + Frostbalance.bot.name + "/" + guildId + "/map.json")
            val gsonBuilder = GsonBuilder()
            gsonBuilder.registerTypeAdapter(TileObject::class.java, TileObjectAdapter())
            val gson = gsonBuilder.create()
            IOUtils.write(file, gson.toJson(map))
        }
    }

    override fun setParent(parent: GameNetwork) {
        while (seed == 0L) {
            seed = Random().nextLong()
        }
        this.gameNetwork = parent
    }
}